### ISSUE TITLE
About low accuracy on converted models

### ISSUE THREAD

**Issue by marcoslucianops** [marcoslucianops/DeepStream-Yolo#339]
I evaluated the mAP between get_wts model and ONNX model and both faced accuracy drop on TensorRT conversion. The conclusion is that the TensorRT drops the accuracy when optimizing the layers.

YOLOv8n ONNX:
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.343
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.492
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.373
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.178
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.381
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.471
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.295
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.488
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.542
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.330
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.599
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.700
```

YOLOv8n get_wts_yolov8.py
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.343
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.491
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.372
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.178
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.381
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.470
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.295
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.488
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.542
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.330
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.599
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.699
```

**Comment by huytranvan2010** [comment]
@marcoslucianops could you please share code to evaluate .engine model?

**Comment by marcoslucianops** [comment]
> @marcoslucianops could you please share code to evaluate .engine model?

I will share it in the future.

**Comment by huytranvan2010** [comment]
> @marcoslucianops could you please share code to evaluate .engine model?

Do I need to use file "libnvdsinfer_custom_impl_Yolo.so" generated from command "CUDA_VER=11.8 make -C nvdsinfer_custom_impl_Yolo" for evaluattion or only use .engine model?

**Comment by marcoslucianops** [comment]
My eval code is created based on [deepstream_python_apps](https://github.com/NVIDIA-AI-IOT/deepstream_python_apps) with some custom implementations (image batch input, pycocotools, etc). It uses DeepStream to generate the JSON to be evaluated by pycocotools.

**Comment by huytranvan2010** [comment]
I inference for each image of COCO val, and collect all of them to generate json file. But I got low mAP for yolov7 fp32 .engine as follow:
```
mAP0.5:0.95 = 0.4
mAP0.5 = 0.538
mAP0.75 = 0.435
```
It is too low compared to your benchmark, even if you use yolov7 fp16 .engine.

**Comment by huytranvan2010** [comment]
> My eval code is created based on [deepstream_python_apps](https://github.com/NVIDIA-AI-IOT/deepstream_python_apps) with some custom implementations (image batch input, pycocotools, etc). It uses DeepStream to generate the JSON to be evaluated by pycocotools.

I inference for each image in COCO val, collect labels to generate json file. But I got low mAP for yolov7 fp32 .engine model:
`
mAP0.5:0.95 = 0.4 
mAP0.5 = 0.538
mAP0.75 = 0.435
`
It is too low compared to your benchmark, even if you use only yolov6 fp16 .engine model

**Comment by marcoslucianops** [comment]
In the models I've tested, there's no mAP difference between FP32 and FP16 engines. Are you using the DeepStream to output the bboxes?

**Comment by huytranvan2010** [comment]
> In the models I've tested, there's no mAP difference between FP32 and FP16 engines. Are you using the DeepStream to output the bboxes?

Yes. I run deepstream app for images and save output (labels) in a file by setting `gie-kitti-output-dir`. Then I collected labels and generated json files to evaluate. My mAP is too low.

**Comment by marcoslucianops** [comment]
In the kitti output, the bboxes coordinates are related to the streammux resolution you set. You need to change them according to each validation image resolution.

**Comment by huytranvan2010** [comment]
> In the kitti output, the bboxes coordinates are related to the streammux resolution you set. You need to change them according to each validation image resolution.

Yes, I recognized that, and also changed to image size, but mAP is too low.

**Comment by marcoslucianops** [comment]
Did you set

```
[class-attrs-all]
nms-iou-threshold=0.65
pre-cluster-threshold=0.001
topk=300
```

In the config_infer_primary_yoloV7.txt file?

**Comment by huytranvan2010** [comment]
> Did you set
> 
> ```
> [class-attrs-all]
> nms-iou-threshold=0.65
> pre-cluster-threshold=0.001
> topk=300
> ```
> 
> In the config_infer_primary_yoloV7.txt file?
Did you use the above config to receive benchmark? I used default set up.
```
nms-iou-threshold=0.45
pre-cluster-threshold=0.25
topk=300
```

**Comment by marcoslucianops** [comment]
The evaluation uses different NMS and confidence thresholds. Try with the values I sent.

**Comment by huytranvan2010** [comment]
> The evaluation uses different NMS and confidence thresholds. Try with the values I sent.

Thanks a lot for supporting me. I am going to try it now😍

**Comment by aidevmin** [comment]
> Did you set
> 
> ```
> [class-attrs-all]
> nms-iou-threshold=0.65
> pre-cluster-threshold=0.001
> topk=300
> ```
> 
> In the config_infer_primary_yoloV7.txt file?

I used this set up, mAP is better, but it is still lower than your benchmark for YOLOv7. Here is my result
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.449
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.623
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.485
```
I attached my config ( I have many config as final_config_1.txt)
[config_infer_primary_yoloV7.txt](https://github.com/marcoslucianops/DeepStream-Yolo/files/11524902/config_infer_primary_yoloV7.txt)
[final_config_1.txt](https://github.com/marcoslucianops/DeepStream-Yolo/files/11524908/final_config_1.txt)

**Comment by huytranvan2010** [comment]
> Did you set
> 
> ```
> [class-attrs-all]
> nms-iou-threshold=0.65
> pre-cluster-threshold=0.001
> topk=300
> ```
> 
> In the config_infer_primary_yoloV7.txt file?

I used this set up, mAP is better, but it is still lower than your benchmark for YOLOv7. Here is my result
```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.449
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.623
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.485
```
I attached my config ( I have many config as final_config_1.txt)
[config_infer_primary_yoloV7.txt](https://github.com/marcoslucianops/DeepStream-Yolo/files/11525057/config_infer_primary_yoloV7.txt)
[final_config_1.txt](https://github.com/marcoslucianops/DeepStream-Yolo/files/11525059/final_config_1.txt)

**Comment by marcoslucianops** [comment]
My eval code is fine-adjusted for extract the better mAP using DeepStream, that's why I got a bit more mAP.

**Comment by huytranvan2010** [comment]
> In the models I've tested, there's no mAP difference between FP32 and FP16 engines. Are you using the DeepStream to output the bboxes?

Do You mean Yolov7 model? I saw that your fp16 .engine model has mAP0.5:0.95 = 0.476, it means that mAP0.5:0.95 (of fp32 .engine model) = 0.476. It is too low compared with reference .pt model mAP0.5:0.95=0.514 https://github.com/WongKinYiu/yolov7#performance

**Comment by huytranvan2010** [comment]
> There's a drop on TensorRT compared to the PyTorch model. In some models, it's a relevant drop. In other models (like PPYOLOE and YOLO-NAS), it's a small. The test I did I was comparing the ONNX export method with the `wts` and `cfg` export method. There's no drop between those two export methods.

Thanks a lot. I expect fp32 is not drop mAP much. If mAP of fp32 or fp16 drop much, so mAP of int8 is still lower.

**Comment by marcoslucianops** [comment]
The FP16 and FP32 mAP are equal.

**Comment by huytranvan2010** [comment]
> The FP16 and FP32 mAP are equal.

Yeah, I think so. In your opinion, what is the reason of  fp32, fp16's mAP big drop compared with .pt models? I mean some models included yolov7. I saw that yolov7 fp16 is dropped about 4%.

**Comment by marcoslucianops** [comment]
In my opinion, TensorRT layers are performance focused, making some tweaks to precisions and parameters. So it's faster, but loses some of the accuracy.

**Comment by cgrtrifork** [comment]
Hi, does this also apply to yoloV8? I noticed a similar issue: after exporting the pytorch model to ONNX and generating the engine file the results were different that when running the pytorch model using the yolo interface directly.

**Comment by cgrtrifork** [comment]
Could this be related to inputs being different, not only TensorRT tweaks? For instance, in YOLOv8 it looks like symmetric padding [is done with a grayscale value](https://github.com/ultralytics/ultralytics/blob/9aaa5d5ed0e5a0c1f053069dd73f12b845c4f282/ultralytics/data/augment.py#L579) rather than with black color [like DeepStream's `nvstreammux` does](https://docs.nvidia.com/metropolis/deepstream/dev-guide/text/DS_plugin_gst-nvstreammux.html#ds-plugin-gst-nvstreammux-properties).

In any case, it would be good to have a table of the expected drop for each of the models, as a reference.

**Comment by WangFengtu1996** [comment]
@cgrtrifork  anything update? I have same warning that Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.  How to slove the problem.

**Comment by WangFengtu1996** [comment]
when I inference yolov8s in Deepstream-6.3 in nvidia agx orin DK, I have some question.
fp32 gpu   ~30fps 
fp16 gpu+dla0 ~11fps  and it's a relevant drop.

would someone give me some explain and guide ?

**Comment by cgrtrifork** [comment]
> I ran the following experiment: I am trying out YOLOv8 object detection on an image that contains an object.
> 
>     1. I used this repository to export the model to onnx. Then using `ffmpeg` I generated a single-frame video, that I feed into DeepStream with a confidence threshold `pre-cluster-threshold=0.2`.
> 
> 
>     * When I use NMS clustering (`cluster-mode=2`, `nms-iou-threshold=0.5`) the object is _not_ found.
> 
>     * If I disable the clustering (`cluster-mode=4`) then an object is found with confidence 0.77.
> 
> 
>     2. I used Triton Inference Server to serve the same TensorRT model that is generated when running DeepStream. Then I ran the inference on the same image. I preprocessed the image to get a 3x640x640 image of float32 between 0 and 1 in RGB format, as it is expected by the model.
> 
> 
>     * When I use gray background for the padding (pixel value = 114/255) —like YOLO does— the max score of the output is 0.87.
> 
>     * When I use black background for the padding (pixel value = 0)—like DeepStream does— the max score of the output is 0.90.
> 
> 
> Having used the same TensorRT model, this makes me think there is an issue either on the parsing and interpretation of the output from the model, or deeper in a lower level DeepStream preprocessing of the image.
> 
>     1. Why does enabling the NMS remove the detection? If the detection is the maximum score the NMS shouldn't remove it.
> 
>     2. Why are the scores different between DeepStream's `nvinfer` plugin and Triton Inference Server?
> 
> 
> For completeness:
> 
>     * The image I'm using was originally extracted from a video by doing:
> 
> 
> ```shell
> # extract all the frames from the original video into a folder
> # frames are enumerated starting from 1
> ffmpeg -i original_video.mp4 original_video/%05d.jpg
> ```
> 
> Then I chose the frame to use (number 84), and I created the single-frame video by doing:
> 
> ```shell
> # frames start from 0, that's why we choose 84-1=83
> ffmpeg -i original_video.mp4 -vf "select=eq(n\,83)" single_frame_video.mp4
> ```
> 
>     * The pipeline I'm using in DeepStream is: `nvurisrcbin` -> `videorate` -> `nvvideoconvert` -> `capsfilter` -> `nvstreammux` -> `queue` -> `nvvideoconvert`  -> `capsfilter` -> `nvinfer` -> `fakesink`. I'm adding a probe after the `nvinfer` to see the detections.
> 
> 
> @marcoslucianops have you tried evaluating the engine file outside of DeepStream?

Following up on this I found out that the parsing from `NvDsInferParseYolo` seems to be correct for this case. However, the resulting detection from DeepStream is not the one with the highest confidence. Here you can see the logs from DeepStream —I added print statements to the library:

```
[Class 0] Box proposal with confidence 0.750208: x1=185.988, y1=141.614, x2=499.038, y2=417.46 (threshold: 0.2)
[Class 0] BBI with confidence 0.750208: left=185.988, top=141.614, width=313.05, height=275.846
[Class 0] Box proposal with confidence 0.881455: x1=184.819, y1=141.771, x2=497.486, y2=416.067 (threshold: 0.2)
[Class 0] BBI with confidence 0.881455: left=184.819, top=141.771, width=312.667, height=274.296
[Class 0] Box proposal with confidence 0.886627: x1=185.479, y1=141.421, x2=499.159, y2=415.547 (threshold: 0.2)
[Class 0] BBI with confidence 0.886627: left=185.479, top=141.421, width=313.68, height=274.127
[Class 0] Box proposal with confidence 0.877862: x1=185.409, y1=141.396, x2=499.173, y2=415.735 (threshold: 0.2)
[Class 0] BBI with confidence 0.877862: left=185.409, top=141.396, width=313.764, height=274.339
[Class 0] Box proposal with confidence 0.866284: x1=184.766, y1=141.94, x2=497.723, y2=416.012 (threshold: 0.2)
[Class 0] BBI with confidence 0.866284: left=184.766, top=141.94, width=312.958, height=274.072
[Class 0] Box proposal with confidence 0.854519: x1=184.601, y1=141.577, x2=499.699, y2=415.742 (threshold: 0.2)
[Class 0] BBI with confidence 0.854519: left=184.601, top=141.577, width=315.097, height=274.165
[Class 0] Box proposal with confidence 0.856617: x1=185.726, y1=141.448, x2=499.246, y2=415.667 (threshold: 0.2)
[Class 0] BBI with confidence 0.856617: left=185.726, top=141.448, width=313.52, height=274.219
[Class 0] Box proposal with confidence 0.770557: x1=184.458, y1=142.046, x2=498.037, y2=416.368 (threshold: 0.2)
[Class 0] BBI with confidence 0.770557: left=184.458, top=142.046, width=313.579, height=274.322
[Class 0] Box proposal with confidence 0.752778: x1=184.416, y1=141.868, x2=499.955, y2=416.512 (threshold: 0.2)
[Class 0] BBI with confidence 0.752778: left=184.416, top=141.868, width=315.539, height=274.644
[Class 0] Box proposal with confidence 0.725658: x1=185.231, y1=141.948, x2=499.762, y2=416.444 (threshold: 0.2)
[Class 0] BBI with confidence 0.725658: left=185.231, top=141.948, width=314.531, height=274.496
[Class 0] Box proposal with confidence 0.23098: x1=184.02, y1=141.643, x2=500.4, y2=416.408 (threshold: 0.2)
[Class 0] BBI with confidence 0.23098: left=184.02, top=141.643, width=316.38, height=274.764
Objects decoded: 11
ObjectList after assignment: 11
2024-02-12 14:21:33,705 [INFO][root]     Frame number: 0
2024-02-12 14:21:33,705 [INFO][root]     [Class 0] Found object with confidence = 0.7705574035644531: left=332.0246276855469, top=0.0827464759349823, width=564.4418334960938, height=494.5528259277344
```
The DeepStream version I'm using is 6.2, I will test this in newer versions too.

**Comment by levipereira** [comment]
Hi @marcoslucianops,
I appreciate your work and for sharing your experiences!
Lately, I've been delving into the precision loss of YOLOv8 models when converted to TensorRT. It's unfortunate that YOLOv8 doesn't natively support ONNX  End2End export.

As my focus lies on model optimization and reducing latencies for edge systems, I've been actively exploring ways to streamline the process. Having struggled with the complexity of the Gst-nvinfer plugin for model construction and the inefficient utilization of GPU resources, I made the switch to Gst-nvinferserver. This allowed me to leverage Triton-Server as a model accelerator, optimizing and instantiating multiple models instances to meet the demands of DeepStream.

However, evaluating accuracy at the TensorRT stage posed a challenge. To address this, I turned to [Triton-Client](https://github.com/levipereira/triton-client-yolo) for evaluating models using the COCO dataset, ensuring end-to-end validation. Surprisingly, I didn't observe significant differences in accuracy between the original PyTorch models and the TensorRT-optimized models for YOLOv7 and YOLOv9. 

Furthermore, I decided against using post-processing plugins for NMS with YOLO models. Instead, I integrated [TRT Efficient NMS](https://github.com/NVIDIA/TensorRT/tree/release/10.0/plugin/efficientNMSPlugin)  directly into the model's final layer. This approach has notably reduced latency by processing NMS on the GPU, thereby minimizing data throughput and eliminating the need for post-processing on the client side. Now, clients only receive the final inference results without requiring additional NMS post-processing.

Another significant advantage for TensorRT users is the utilization of QAT models instead PQT. QAT models exhibit minimal accuracy loss while offering significant optimization in inference per second and latency reduction.

When you have a moment, take a look at this repository:
https://github.com/levipereira/triton-server-yolo
https://github.com/levipereira/deepstream-yolo-triton-server-rtsp-out

**Comment by marcoslucianops** [comment]
The issue is related to the DeepStream pipeline. The way it does processing reduces the accuracy of the detections.

**Comment by Zanz2** [comment]
We are using a ultralytics yolov11 and while we havent done any comprehensive analysis yet it seems like it performs much worse than its counterpart (ultralytics yolov5) on the sample720p.h264 file. We see a lot of "blinking" on certain objects in corners (that are not occluded) and also swings in confidence scores.

You mention that this is due to deepstreams optimizations, so i assume there would be no way to fix it with the plugin, since the internals of deepstream are a black box.

**Comment by marcoslucianops** [comment]
The decrease in the accuracy is related to the way NVIDIA does pre-processing in the pipeline elements. It's not related to the YOLO plugin. I'm able to get accuracy in pair to Torch model, but it's something I use only in my commercial projects.

**Comment by marcoslucianops** [comment]
One step to confirm the accuracy with this repo is using the YOLO plugin with TensorRT python code. It's possible to get same accuracy of Torch model using that way.

**Comment by Nabyssache** [comment]
Regarding the issue of accuracy reduction, I have a few questions that I would like to ask you:
1. When deepstream's pipline reasoning H.264/265 online stream, the picture will automatically resize to 1920*1080. Is this the key reason affecting the accuracy? When exporting onnx,  need to input w/h to a scale of 1920*1080?
2. The yolov11 training enables AMP. When exporting,  need to directly switch to half half precision?
3. Will the onnx export parameter --simplify affect the accuracy?
The above is the experimental direction I thought of. Trying to do some experiments to solve the accuracy problem. Yolov11 often has inexplicable false positives in actual projects with deepstream, so it has to be replaced with a PT model and code. Have you tried these directions? （So I don't have to repeatedly stumble）

**Comment by Nabyssache** [comment]
> One step to confirm the accuracy with this repo is using the YOLO plugin with TensorRT python code. It's possible to get same accuracy of Torch model using that way.

Regarding the issue of accuracy reduction, I have a few questions that I would like to ask you:

When deepstream's pipline reasoning H.264/265 online stream, the picture will automatically resize to 19201080. Is this the key reason affecting the accuracy? When exporting onnx, need to input w/h to a scale of 19201080?
The yolov11 training enables AMP. When exporting, need to directly switch to half half precision?
Will the onnx export parameter --simplify affect the accuracy?
The above is the experimental direction I thought of. Trying to do some experiments to solve the accuracy problem. Yolov11 often has inexplicable false positives in actual projects with deepstream, so it has to be replaced with a PT model and code. Have you tried these directions? （So I don't have to repeatedly stumble）

@marcoslucianops

**Comment by marcoslucianops** [comment]
Yes, the issue is the pre-processing/decoding (changes in the frame before the inference) in DeepStream. With the default pipeline, it isn't possible to avoid that.

We don't need to export the model in 1920x1088, you should use the same training resolution.

About the AMP, we don't need to set to half when exporting.

`--simplify` doesn't affect the accuracy.

**Comment by tamil-03** [comment]
> Another report: We are using a ultralytics yolov11 and while we havent done any comprehensive analysis yet it seems like it performs much worse than its counterpart (ultralytics yolov5) on the sample720p.h264 file. We see a lot of "blinking" on certain objects in corners (that are not occluded) and also swings in confidence scores.
> 
> You mention that this is due to deepstreams optimizations, so i assume there would be no way to fix it with the plugin, since the internals of deepstream are a black box. But still very shocking, because the difference between running the same file in python is so stark.

Hi @Zanz2 ,
I faced the same issue you mentioned, the blinking effect on objects near the corners. In my case, adding **crop-objects-to-roi-boundary=1** under the [property] section of the nvinfer model config file resolved the problem. Please check this configuration; it might be causing the same issue on your end.

Hi @marcoslucianops, 
The issue I mentioned above could be one of the reasons for the observed accuracy drop. However, this behavior applies only to DeepStream, as I verified from the NvInfer source code. NvInfer discards detections that extend beyond the frame dimensions if **crop-objects-to-roi-boundary** is not set to 1. When running inference using the TensorRT C++ API, I did not encounter this issue.

In /opt/nvidia/deepstream/deepstream/sources/gst-plugins/gst-nvinfer/gstnvinfer_meta_utils.cpp, check the lines 103 and 113.

<img width="1102" height="624" alt="Image" src="https://github.com/user-attachments/assets/03d7ba64-d927-4435-abd8-eef7d58b2d6b" />
